# EXP 2.2    
### wine data classification


In [1]:
# 필요한 모듈 import 하기
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

# 데이터 준비
wine = load_wine()         


wine_data = wine.data             #feature data 지정
                                      #wine_data라는 변수에 wine의 데이터 입력

wine_label = wine.target          #label data 지정
                                      #wine의 label 정보를 변수에 저장

print('"target name은 %s 입니다."' %wine.target_names)            #target names 출력  
print("\n")                                                       #wine의 종류 알 수 있다 


# train, test 데이터 분리 
# 모델 학습과 테스트용 문제지와 정답지 준비 
X_train, X_test, y_train, y_test = train_test_split (wine_data,      # 모델의 feature 
                                                    wine_label,      # 모델이 맞춰야 하는 정답 (0-9까지의 숫자)
                                                    test_size = 0.2,   # 전체 데이터 중 20%를 test데이터로 지정
                                                    random_state=7)    # 랜덤으로 train/test data 분리 


"target name은 ['class_0' 'class_1' 'class_2'] 입니다."




In [2]:
# target name이 class_0,1,2로 되어 있어 어떤 특징으로 분류 되었는지 알기 어렵다
# 와인 데이터 분류 특징/기준

wine.feature_names

['alcohol',
 'malic_acid',
 'ash',
 'alcalinity_of_ash',
 'magnesium',
 'total_phenols',
 'flavanoids',
 'nonflavanoid_phenols',
 'proanthocyanins',
 'color_intensity',
 'hue',
 'od280/od315_of_diluted_wines',
 'proline']

위와 같은 여러 특징을 통하여 와인을 3개의 class로 분류하는 것을 알 수 있다.

---
### 1) Decision Tree 모델

In [3]:
from sklearn.tree import DecisionTreeClassifier

#(5) 모델 학습 - X_train, y_train 활용하여 분류기 모델 만들기 / 성능 체크
decision_tree = DecisionTreeClassifier(random_state = 32)
decision_tree.fit(X_train,y_train)      # 모델학습 - training dataset에 맞게 패턴 파악 및 예측/학습
y_pred= decision_tree.predict(X_test)

print(classification_report(y_test,y_pred))

#성능평가
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test,y_pred)
accuracy

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       0.89      1.00      0.94        17
           2       1.00      0.83      0.91        12

    accuracy                           0.94        36
   macro avg       0.96      0.94      0.95        36
weighted avg       0.95      0.94      0.94        36



0.9444444444444444

---
### 2) Random Forest 모델

In [11]:
from sklearn.ensemble import RandomForestClassifier


random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))


from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test,y_pred)
accuracy


              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       1.00      1.00      1.00        17
           2       1.00      1.00      1.00        12

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36



1.0

---
### 3) SVM 모델

In [5]:
from sklearn import svm
svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test,y_pred)
accuracy


              precision    recall  f1-score   support

           0       0.86      0.86      0.86         7
           1       0.58      0.88      0.70        17
           2       0.33      0.08      0.13        12

    accuracy                           0.61        36
   macro avg       0.59      0.61      0.56        36
weighted avg       0.55      0.61      0.54        36



0.6111111111111112

---
### 4) SGD Classifier 모델

In [8]:
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
accuracy

              precision    recall  f1-score   support

           0       0.50      1.00      0.67         7
           1       0.88      0.41      0.56        17
           2       0.43      0.50      0.46        12

    accuracy                           0.56        36
   macro avg       0.60      0.64      0.56        36
weighted avg       0.65      0.56      0.55        36



0.5555555555555556

---
### 5) Logistic Regression 모델

In [9]:
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()
# 오류나서 추가함 
logistic_model = LogisticRegression(solver='lbfgs', max_iter=10000).fit(X_train, y_train)
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)

accuracy


              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       0.94      1.00      0.97        17
           2       1.00      0.92      0.96        12

    accuracy                           0.97        36
   macro avg       0.98      0.97      0.98        36
weighted avg       0.97      0.97      0.97        36



0.9722222222222222

---
# 분석하기 
정답(label)의 분포가 7-36으로 불균형하다고 판단하여 accuracy 말고 다른 모델로 선택이 필요하다.  
다양한 특징을 통하여 traget class로 분류 되어야하므로 , 와인의 정확한 특징을 놓치지 않고 선별하는 과정이 필요할 것으로 보인다.  
따라서 특징을 가진다고 분류 된 와인이 실제로 특징을 가지는 것의 확률이 높아야 하며,  
특징을 가지고 있지 않는데 가지고 있다고 잘못 판단되는 상황이 나타나면 고객이 구매할 때 어려움을 느낄 수 있다!  
  
  ∴ **Precision 이 중요하다 !!**


- 5가지 모델의 Precision 비교 
  - Decision Tree   : 0.96  
  - Random Forest: 1.0
  - SVM : 0.59
  - SGD Classifier : 0.6
  - Logistic Regression : 0.98  
  
Precision의 비교를 통하여 **Random Forest** 모델이 1.0으로 가장 높게 나왔다.  
 
 